### PR TITLE
Dont keeping key_list in transaction

### PR DESCRIPTION
**Pull request by xiaoxichen**
The RocksDB/LevelDB  PUT/DELETE api already make a deep copy of key/value, we don't need to maintain a copy outside.

### CODE REVIEW

**Comment by xiaoxichen** [comment]
@yuyuyu101 , could you pls take a review?

**Comment by yuyuyu101** [comment]
LevelDB' slice structure won't maintain a copy of data, so we need to keep the bufferptr reference at LevelDBStore/RocksDBStore.

Plz correct me if wrong.

**Comment by xiaoxichen** [comment]
@yuyuyu101 , partially true:)

Slice structure won't maintain a copy, that's true, But, the PUT/DELETE API will do the copy , see

https://github.com/xiaoxichen/leveldb/blob/master/db/write_batch.cc#L101

void WriteBatch::Put(const Slice& key, const Slice& value) {
 WriteBatchInternal::SetCount(this, WriteBatchInternal::Count(this) + 1);
  rep_.push_back(static_cast<char>(kTypeValue));
  PutLengthPrefixedSlice(&rep_, key);
  PutLengthPrefixedSlice(&rep_, value);
}

and 
https://github.com/xiaoxichen/leveldb/blob/master/util/coding.cc#L98

void PutLengthPrefixedSlice(std::string* dst, const Slice& value) {
  PutVarint32(dst, value.size());
  dst->append(value.data(), value.size());
}

**Comment by xiaoxichen** [comment]
@yuyuyu101 , partially true:)

Slice structure won't maintain a copy, that's true, But, the PUT/DELETE API will do the copy , see

https://github.com/xiaoxichen/leveldb/blob/master/db/write_batch.cc#L101

void WriteBatch::Put(const Slice& key, const Slice& value) {
 WriteBatchInternal::SetCount(this, WriteBatchInternal::Count(this) + 1);
  rep_.push_back(static_cast<char>(kTypeValue));
  PutLengthPrefixedSlice(&rep_, key);
  PutLengthPrefixedSlice(&rep_, value);
}

and 
https://github.com/xiaoxichen/leveldb/blob/master/util/coding.cc#L98

void PutLengthPrefixedSlice(std::string* dst, const Slice& value) {
  PutVarint32(dst, value.size());
  dst->append(value.data(), value.size());
}

**Comment by xiaoxichen** [comment]
similar code applied in RocksDB

**Comment by yuyuyu101** [comment]
Reviewed-by:

@liewegas plz take a look, it seemed helpful to solve radosgw memory case(use omap heavily)

File: src/os/LevelDBStore.cc [hunk 1]
```diff
@@ -162,21 +162,19 @@ void LevelDBStore::LevelDBTransactionImpl::set(
...
   string key = combine_strings(prefix, k);
-  keys.push_back(key);
-  bat.Delete(leveldb::Slice(*(keys.rbegin())));
-  bat.Put(leveldb::Slice(*(keys.rbegin())),
-	  leveldb::Slice(bl.c_str(), bl.length()));
+  //bufferlist::c_str() is non-constant, so we need to make a copy
+  bufferlist val = to_set_bl;
+  bat.Delete(leveldb::Slice(key));
+  bat.Put(leveldb::Slice(key),
+	  leveldb::Slice(val.c_str(), val.length()));
```
Comment (varadakari) [inline src/os/LevelDBStore.cc]: can we change c_str() to data(), we can avoid one more data copy.

Comment (varadakari) [inline src/os/LevelDBStore.cc]: please ignore my previous comment. thought that was a string.

File: src/os/LevelDBStore.cc [hunk 2]
```diff
@@ -162,21 +162,19 @@ void LevelDBStore::LevelDBTransactionImpl::set(
...
 }
 
 void LevelDBStore::LevelDBTransactionImpl::rmkey(const string &prefix,
 					         const string &k)
 {
   string key = combine_strings(prefix, k);
-  keys.push_back(key);
-  bat.Delete(leveldb::Slice(*(keys.rbegin())));
+  bat.Delete(leveldb::Slice(key));
```
Comment (varadakari) [inline src/os/LevelDBStore.cc]: Here we are having taking only a reference, don't we need the data pointer of the key? key goes out of scope after this function.

Comment (xiaoxichen) [inline src/os/LevelDBStore.cc]: no need, the leveldb api will do a memcpy inside the api by themselves

**Comment by varadakari** [comment]
LGTM

File: src/os/LevelDBStore.cc (hunk 1 above)
Comment (liewegas) [inline src/os/LevelDBStore.cc]: we should make a data() accessor for bufferlist etc .. the c_str() inconsistency is confusing

**Comment by liewegas** [comment]
Looks good!

**Comment by tchaikov** [comment]
tested at
- http://pulpito.ceph.com/ubuntu-2015-06-29_18:58:24-rados-wip-kefu-testing---basic-multi/
- http://pulpito.ceph.com/ubuntu-2015-06-30_20:30:57-rados-wip-kefu-testing---basic-multi/

the failed ones are tracked at http://tracker.ceph.com/issues/12192, which are also spotted on the master branch.

**Pull request merged**